### PR TITLE
tests: Skip SQLite fsyncs while testing

### LINKED ISSUE #21628: Up to 100x perfomance difference for sqlite vs BDB in CreateWallet test
Building master @ 41a8d2b96ff5cbc2efa9b0996f30228d0a625fdb (sqlite):
```bash
./autogen.sh
./configure --without-bdb --without-gui --with-sqlite
  with wallet     = yes
    with sqlite   = yes
    with bdb      = no
gmake -j8
src/test/test_bitcoin --run_test=wallet_tests --log_level=all
...
wallet/test/wallet_tests.cpp:696: Entering test case "CreateWallet"
wallet/test/wallet_tests.cpp:69: info: check SignTransaction(mtx, &keystore, coins, SIGHASH_ALL, input_errors) has passed
wallet/test/wallet_tests.cpp:69: info: check SignTransaction(mtx, &keystore, coins, SIGHASH_ALL, input_errors) has passed
wallet/test/wallet_tests.cpp:734: info: check m_node.chain->broadcastTransaction(MakeTransactionRef(mempool_tx), DEFAULT_TRANSACTION_MAXFEE, false, error) has passed
wallet/test/wallet_tests.cpp:740: info: check rescan_completed has passed
wallet/test/wallet_tests.cpp:741: info: check addtx_count == 2 has passed
wallet/test/wallet_tests.cpp:744: info: check wallet->mapWallet.count(block_tx.GetHash()) == 1U has passed
wallet/test/wallet_tests.cpp:745: info: check wallet->mapWallet.count(mempool_tx.GetHash()) == 1U has passed
wallet/test/wallet_tests.cpp:753: info: check addtx_count == 4 has passed
wallet/test/wallet_tests.cpp:767: info: check rescan_completed has passed
wallet/test/wallet_tests.cpp:69: info: check SignTransaction(mtx, &keystore, coins, SIGHASH_ALL, input_errors) has passed
wallet/test/wallet_tests.cpp:69: info: check SignTransaction(mtx, &keystore, coins, SIGHASH_ALL, input_errors) has passed
wallet/test/wallet_tests.cpp:772: info: check m_node.chain->broadcastTransaction(MakeTransactionRef(mempool_tx), DEFAULT_TRANSACTION_MAXFEE, false, error) has passed
wallet/test/wallet_tests.cpp:780: info: check addtx_count == 4 has passed
wallet/test/wallet_tests.cpp:783: info: check wallet->mapWallet.count(block_tx.GetHash()) == 1U has passed
wallet/test/wallet_tests.cpp:784: info: check wallet->mapWallet.count(mempool_tx.GetHash()) == 1U has passed
wallet/test/wallet_tests.cpp:696: Leaving test case "CreateWallet"; testing time: 201296644us
wallet/test/wallet_tests.cpp:791: Entering test case "ZapSelectTx"
wallet/test/wallet_tests.cpp:69: info: check SignTransaction(mtx, &keystore, coins, SIGHASH_ALL, input_errors) has passed
wallet/test/wallet_tests.cpp:810: info: check wallet->HasWalletSpend(prev_hash) has passed
wallet/test/wallet_tests.cpp:811: info: check wallet->mapWallet.count(block_hash) == 1u has passed
wallet/test/wallet_tests.cpp:814: info: check wallet->ZapSelectTx(vHashIn, vHashOut) == DBErrors::LOAD_OK has passed
wallet/test/wallet_tests.cpp:816: info: check !wallet->HasWalletSpend(prev_hash) has passed
wallet/test/wallet_tests.cpp:817: info: check wallet->mapWallet.count(block_hash) == 0u has passed
wallet/test/wallet_tests.cpp:791: Leaving test case "ZapSelectTx"; testing time: 200866629us
```

Compare that to running the same tests after building with just bdb:
```bash
gmake clean
./configure --without-gui --with-sqlite=no
  with wallet     = yes
    with sqlite   = no
    with bdb      = yes
gmake -j8
src/test/test_bitcoin --run_test=wallet_tests --log_level=all
...
wallet/test/wallet_tests.cpp:696: Entering test case "CreateWallet"
wallet/test/wallet_tests.cpp:69: info: check SignTransaction(mtx, &keystore, coins, SIGHASH_ALL, input_errors) has passed
wallet/test/wallet_tests.cpp:69: info: check SignTransaction(mtx, &keystore, coins, SIGHASH_ALL, input_errors) has passed
wallet/test/wallet_tests.cpp:734: info: check m_node.chain->broadcastTransaction(MakeTransactionRef(mempool_tx), DEFAULT_TRANSACTION_MAXFEE, false, error) has passed
wallet/test/wallet_tests.cpp:740: info: check rescan_completed has passed
wallet/test/wallet_tests.cpp:741: info: check addtx_count == 2 has passed
wallet/test/wallet_tests.cpp:744: info: check wallet->mapWallet.count(block_tx.GetHash()) == 1U has passed
wallet/test/wallet_tests.cpp:745: info: check wallet->mapWallet.count(mempool_tx.GetHash()) == 1U has passed
wallet/test/wallet_tests.cpp:753: info: check addtx_count == 4 has passed
wallet/test/wallet_tests.cpp:767: info: check rescan_completed has passed
wallet/test/wallet_tests.cpp:69: info: check SignTransaction(mtx, &keystore, coins, SIGHASH_ALL, input_errors) has passed
wallet/test/wallet_tests.cpp:69: info: check SignTransaction(mtx, &keystore, coins, SIGHASH_ALL, input_errors) has passed
wallet/test/wallet_tests.cpp:772: info: check m_node.chain->broadcastTransaction(MakeTransactionRef(mempool_tx), DEFAULT_TRANSACTION_MAXFEE, false, error) has passed
wallet/test/wallet_tests.cpp:780: info: check addtx_count == 4 has passed
wallet/test/wallet_tests.cpp:783: info: check wallet->mapWallet.count(block_tx.GetHash()) == 1U has passed
wallet/test/wallet_tests.cpp:784: info: check wallet->mapWallet.count(mempool_tx.GetHash()) == 1U has passed
wallet/test/wallet_tests.cpp:696: Leaving test case "CreateWallet"; testing time: 2139054us
wallet/test/wallet_tests.cpp:791: Entering test case "ZapSelectTx"
wallet/test/wallet_tests.cpp:69: info: check SignTransaction(mtx, &keystore, coins, SIGHASH_ALL, input_errors) has passed
wallet/test/wallet_tests.cpp:810: info: check wallet->HasWalletSpend(prev_hash) has passed
wallet/test/wallet_tests.cpp:811: info: check wallet->mapWallet.count(block_hash) == 1u has passed
wallet/test/wallet_tests.cpp:814: info: check wallet->ZapSelectTx(vHashIn, vHashOut) == DBErrors::LOAD_OK has passed
wallet/test/wallet_tests.cpp:816: info: check !wallet->HasWalletSpend(prev_hash) has passed
wallet/test/wallet_tests.cpp:817: info: check wallet->mapWallet.count(block_hash) == 0u has passed
wallet/test/wallet_tests.cpp:791: Leaving test case "ZapSelectTx"; testing time: 1821042us
```

In the sqlite case,`CreateWallet` is taking ~100x longer: `201296644us` vs `2139054us`. `ZapWalletTx` is similar: `200866629us` vs `1821042us`. [Full logs here](https://gist.github.com/fanquake/09e983acf0792482534f0fe7e916a03a).

I thought this might be some weird macOS behaviour, but I tested on Linux, and am also seeing up to 27x difference. `CreateWallet`: `27905993us` vs `1092394us`. `ZapWalletTx`: `28024310us` vs `969432us`. [Full logs here](https://gist.github.com/fanquake/a1104888b5574af00a40a9998cfd0b36).

### PR DESCRIPTION
**Pull request by achow101**
Since we want tests to run quickly, and since tests do a lot more db operations than expected we expect to see in actual usage, we disable sqlite's syncing behavior to make db operations run much faster. This syncing behavior is necessary for normal operation as it helps guarantee that data won't become lost or corrupted, but in tests, we don't care about that.

Fixes #21628

### CODE REVIEW

File: src/wallet/sqlite.cpp [hunk 1]
```diff
@@ -233,6 +233,15 @@ void SQLiteDatabase::Open()
         throw std::runtime_error(strprintf("SQLiteDatabase: Failed to enable fullfsync: %s\n", sqlite3_errstr(ret)));
     }
 
+    if (gArgs.GetBoolArg("-unsafesqlitesync", false)) {
+        // Use normal synchronous mode for the journal
+        LogPrintf("SQLiteDatabase: WARNING SQLite is configured to not wait for data to be flushed to disk. Data loss and corruption may occur.");
```
Comment (MarcoFalke) [inline src/wallet/sqlite.cpp:238]:
```suggestion
        LogPrintf("WARNING SQLite is configured to not wait for data to be flushed to disk. Data loss and corruption may occur.\n");
```

function names and file paths are already logged in tests. Also missing `\n` to make linter happy?

**Review by maflcko — COMMENTED** [review]
Left a linter comment, didn't review

File: src/wallet/sqlite.cpp (hunk 1 above)
Comment (achow101) [reply src/wallet/sqlite.cpp:238]: Done

File: src/wallet/init.cpp [hunk 2]
```diff
@@ -82,6 +82,12 @@ void WalletInit::AddWalletOptions(ArgsManager& argsman) const
     argsman.AddHiddenArgs({"-dblogsize", "-flushwallet", "-privdb"});
 #endif
 
+#ifdef USE_SQLITE
+    argsman.AddArg("-unsafesqlitesync", "Set SQLite synchronous=OFF to disable waiting for the database to sync to disk. This is unsafe and can cause data loss and corruption. This option is only used by tests to improve their performance (default: false)", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::WALLET_DEBUG_TEST);
```
Comment (vasild) [inline src/wallet/init.cpp:86]: Since this is a boolean argument, why not `s/ArgsManager::ALLOW_ANY/ArgsManager::ALLOW_BOOL/`

**Review by vasild — COMMENTED** [review]
How much performance boost does this bring?

File: test/functional/test_framework/test_node.py [hunk 3]
```diff
@@ -103,6 +103,8 @@ def __init__(self, i, datadir, *, chain, rpchost, timewait, timeout_factor, bitc
             "-debugexclude=leveldb",
             "-uacomment=testnode%d" % i,
         ]
+        # To improve SQLite wallet performance so that the tests don't timeout, use -unsafesqlitesync
+        append_config(datadir, ["unsafesqlitesync=1"])
```
[the remark this answers is not in the document]
Comment (achow101) [reply test/functional/test_framework/test_node.py:107]: `self.args` is command line only, not in the conf file. I wanted this to be in the bitcoin.conf file, although thinking on that, I'm not sure if it is necessary.

File: src/wallet/init.cpp (hunk 2 above)
Comment (achow101) [reply src/wallet/init.cpp:86]: Done

**Review by vasild — APPROVED** [review]
ACK 5c72335f0f848895079913e58534ac9488c18ed7

Locally,`wallet_tests/CreateWallet` test time [seconds]:

```
master (cb79cabd) sqlite: 63, 62, 60
master (cb79cabd) bdb:     2,  3,  3
PR     (5c72335f) sqlite:  3,  3,  3
```

Maybe [add `-unsafesqlitesync=1` to `self.args[]`](https://github.com/bitcoin/bitcoin/pull/21634#discussion_r610726804) if there is no reason to use a separate call to `append_config()`.

File: test/functional/test_framework/test_node.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vasild) [reply test/functional/test_framework/test_node.py:107]: I think this question will rise every time somebody looks here (even after merge). Maybe move it to `self.args[]`, or if there is a reason to use `append_config()`, then mention that reason in a comment.

**Comment by laanwj** [comment]
Concept ACK. Syncing is a robustness measure, and unless you're testing the syncing itself, there is no use to have it enabled in tests.

Seperately it would make sense to look into the sqlite performance issues, but this is a win.

File: test/functional/test_framework/test_node.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply test/functional/test_framework/test_node.py:107]: It is possible to modify `write_config` if the goal is to add a static option to the conf file

[the remark this answers is not in the document]
Comment (achow101) [reply test/functional/test_framework/test_node.py]: `write_config` was the function I was looking for. Changed to use that instead.

**Review by vasild — APPROVED** [review]
ACK 41f891da508114f1fd4df30b4068073ec30abc2a

**Pull request merged**